### DISCUSSION TITLE
SearchBar text prop

### DISCUSSION THREAD

**Discussion by salisuabubakar** [grahammendick/navigation#676] [Q&A] [answered]
Hello, Thank you very much for the wonderfull job you are doing with The Navigation Router. I have used Wix RNN and RN but i can confidently tell that by far The Navigation Router is the best interms of performace. I was very very impressed. I moved an app from Wix RNN to The Navigation Router and the performance was about 30 percent.

I have an issue with the SearchBar. Can you explain wha the text prop is used for and if possible can you share a code with example of how to use it?  Am very new to React Native.  Thanks

          <SearchBar
            **text={text}**
            autoCapitalize="none"
            obscureBackground={false}
            barTintColor={() => 'white'}
            onChangeText={text => onChangeText(text)}>
            <ScrollView></ScrollView>
          </SearchBar>
        </NavigationBar>

**Comment by grahammendick** [comment, marked as answer]
Hey @salisuabubakar. Thank you so much. You’ve really put a smile on my face. I’m surprised you only saw a 30% improvement ;)

The text prop works the same as the [value prop on the React Native TextInput component.](https://reactnative.dev/docs/textinput) You need `useState` to hold the text.
```jsx
const [text, onChangeText] = useState('');

<SearchBar
  text={text}
  onChangeText={onChangeText}>
</SearchBar>
```

There’s a [zoom sample that shows the SearchBar](https://github.com/grahammendick/navigation/blob/0be50babe33c042ffc2e19d3896ebbf9b973156a/NavigationReactNative/sample/zoom/Grid.js#L41) in action. I’ve recently updated the `SearchBar` component to support the [Material 3 Search Bar from android](https://m3.material.io/components/search/overview). If you want to take advantage of this, like in the zoom sample, then you should update to the latest navigation-react-native and use a Material 3 Theme.

Any questions let me know. I’m always happy to help.

**Reply by salisuabubakar** [reply]
Hi @grahammendick ,
Thanks very much for your quick response. Am surprise you respond so quickly. Like i mentioned earlier, am new to react native and the 30% improvement based on the migration i did but am quite sure it will be more than 30% when i finish migration. I can feel i am using a Native App becasuse everything is smooth The Navigation Router.

I really need your help with the search. I have tried everything. I was using input component and it works but using the Search component in The Navigation Router it doesnt work. I know this is too much to ask but after getting it to work will help with the migration.

Below is the code am using. 
```jsx
 const [filterProductSearchData, setfilterProductSearchData] = useState([]);
  const [productSearchData, setproductSearchData] = useState([]);
  const [featprogramData, setfeatprogramData] = useState([]);   

  const fetchFeatureprogram = () => {
    return axios.get('https://example.com/data.json')
      .then(data => {
        const resData = data.data;
        setfeatprogramData(resData);
        // console.log(JSON.stringify(resData))
      })
      .catch(error => {
        console.log(error);
      })
  }

   const onChangeText = text => {
    let productDataFilter = filterProductSearchData;
    let productSearchResult = productDataFilter.filter(
      itemData =>
        itemData.title.includes(text) || itemData.price__number.includes(text),
    );
    setproductSearchData(productSearchResult);
  };

<SafeAreaView style={styles.safeAreaView}>
      <ScrollView contentInsetAdjustmentBehavior="automatic">
        <NavigationBar
          title="Programs"
          largeTitle={true}
          titleColor={'#ffffff'}
          barTintColor={'#cd3527'}
          navigationAccessibilityLabel="Back"
          tintColor={"#ffffff"}
          onNavigationPress={() => { stateNavigator.navigateBack(1) }}
        >
          <StatusBar barTintColor={'#cd3527'} tintStyle="light" hidden={false} />
          <SearchBar
            text={text}
            autoCapitalize="none"
            obscureBackground={false}
            barTintColor={() => 'white'}
            onChangeText={text => onChangeText(text)}
            hideNavigationBar={false}
            hideWhenScrolling={false}
          >
            <ScrollView></ScrollView>
          </SearchBar>
        </NavigationBar>
        <FlatList
          data={featprogramData}
          ItemSeparatorComponent={space}
          keyExtractor={() => Math.random().toString(12).substring(0)}
          ListEmptyComponent={<IsLoadingBockPages />}
          horizontal={false}
          showsVerticalScrollIndicator={false}
          showsHorizontalScrollIndicator={false}
          automaticallyAdjustContentInsets={false}
          contentInsetAdjustmentBehavior="automatic"
          initialNumToRender={5}
          centerContent={true}
          contentContainerStyle={{ padding: 16, flexGrow: 1 }}
          renderItem={renderFeaturedprogram}
        />
      </ScrollView>
    </SafeAreaView>
```

Thanks for your time and the wonderful library. I wish you have a donation link where people can donate some token for your effort.

**Reply by grahammendick** [reply]
Hi, I'm happy to help. There's a few things not quite right with your code sample

1. Don't wrap the screen in a `SafeAreaView`. Instead, create a `Container` component that renders a `ScrollView` on iOS and a `View` on Android. Then wrap the `NavigationBar` and `FlatList` in only this `Container`.
 ```jsx
const Container = (props) => (
  Platform.OS === 'ios'
    ? <ScrollView contentInsetAdjustmentBehavior="automatic" {...props}/>
    : <View collapsable={false} {...props} />
);

<Container>
  <NavigationBar>
  <FlatList>
</Container>
```
2. Render your search results inside the `ScrollView` child of the `SearchBar`. This `ScrollView` must also have  `contentInsetAdjustmentBehavior` set to 'automatic'.
```jsx
<SearchBar>
  <ScrollView contentInsetAdjustmentBehavior="automatic">
    <View>
      {productSearchData.map(() => )}
    </View>
  </ScrollView>
</SearchBar>
```
3. Store the text in `useState` and do your search when the text changes
```jsx
const [text, setText] = useState('');
const onChangeText = text => {
  setText(text);
  ...
  setproductSearchData(productSearchResult);
};

<SearchBar text={text} onChangeText={onChangeText}>
```

If it's still now working after these changes, please let me know exactly what's going wrong and whether it's on iOS, Android or both. Maybe include a video if you can?

Also, I really recommend [running the zoom sample](https://github.com/grahammendick/navigation/tree/master/NavigationReactNative/sample/zoom). It has a [Grid component with a SearchBar](https://github.com/grahammendick/navigation/blob/98b4494479d638f3f459dad538cfc4733602a29f/NavigationReactNative/sample/zoom/Grid.js#L44).

**Reply by grahammendick** [reply]
I made a couple of edits to the code samples above

**Reply by salisuabubakar** [reply]
Hi @grahammendick ,
Thanks once again for the quick response. I have tried all i can to make sure this is working. I really want to get it working and i appreciate your time. I have attached the full code below.  I believe if you paste it you will get the full component working.

```jsx
import React, { useState, useEffect, useLayoutEffect } from 'react';
import {
  View,
  StyleSheet,
  Text,
  FlatList,
  ScrollView,
  Platform
} from 'react-native';
import AnimatedImageLoader from 'lottie-react-native';
import FastImage from 'react-native-fast-image';
import axios from 'axios';
import { NavigationBar, StatusBar, useNavigated, useUnloaded, SearchBar } from 'navigation-react-native';



function ProgramsScreen() {
  const [filterProductSearchData, setfilterProductSearchData] = useState([]);
  const [productSearchData, setproductSearchData] = useState([]);
  const [featprogramData, setfeatprogramData] = useState([]);
  const [text, setText] = useState('');


  const space = () => {
    return <View style={{ height: 30, backgroundColor: '#ffffff' }} />
  };


  const IsLoadingBockPages = () => {
    return (
      <View style={styles.loadingEmptyContainer}>
        <AnimatedImageLoader
          source={require('../../assets/loaderView.json')}
          autoPlay={true}
          loop={true}
          autoSize={true}
          style={styles.loaderView}
        />
      </View>
    );
  };


  useEffect(() => {
    fetchFeatureprogram();
  }, []);


  useNavigated(() => {
    fetchFeatureprogram();
    console.log('ComponentDidAppear',);
  })

  useUnloaded(() => {
    fetchFeatureprogram();
    console.log('ComponentDidDisappear');
  })

  const fetchFeatureprogram = () => {
    return axios.get('https://example.com/programms.json')
      .then(data => {
        const resData = data.data;
        setfeatprogramData(resData);
        // console.log(JSON.stringify(resData))
      })
      .catch(error => {
        console.log(error);
      })
  }


  const renderFeaturedprogram = ({ item }) => {
    return (
      <View style={styles.progWrapper} key={item.id}>
        <FastImage
          style={styles.bars}
          source={require('../../assets/images/bars.png')}
          resizeMode={FastImage.resizeMode.contain}
          tintColor={'orange'}
        />
        <View>
          <Text adjustsFontSizeToFit={true} allowFontScaling style={styles.programtitle}>{item.programtitle}</Text>
          <Text adjustsFontSizeToFit={true} allowFontScaling style={styles.programtime}>{item.programtime}</Text>
        </View>
      </View>
    )
  }


  const onChangeText = text => {
    setText(text);
    let productDataFilter = filterProductSearchData;
    let productSearchResult = productDataFilter.filter(
      itemData =>
        itemData.programtitle.includes(text) || itemData.programtime.includes(text),
    ); setproductSearchData(productSearchResult);
  };


  const Container = (props) => (
    Platform.OS === 'ios'
      ? <ScrollView contentInsetAdjustmentBehavior="automatic" {...props} />
      : <View collapsable={false} {...props} />
  );


  return (
    <Container>
      <NavigationBar
        title="Programs"
        largeTitle={true}
        titleColor={'#ffffff'}
        barTintColor={'#cd3527'}
        navigationAccessibilityLabel="Back"
        tintColor={"#ffffff"}
        onNavigationPress={() => { stateNavigator.navigateBack(1) }}
      >
        <StatusBar barTintColor={'#cd3527'} tintStyle="light" hidden={false} />
        <SearchBar
          text={""}
          autoCapitalize="none"
          obscureBackground={false}
          barTintColor={() => 'white'}
          onChangeText={text => onChangeText(text)}
          hideNavigationBar={false}
          hideWhenScrolling={false}
        >
          <ScrollView></ScrollView>
        </SearchBar>
      </NavigationBar>
      <FlatList
        data={featprogramData}
        ItemSeparatorComponent={space}
        keyExtractor={() => Math.random().toString(12).substring(0)}
        ListEmptyComponent={<IsLoadingBockPages />}
        horizontal={false}
        showsVerticalScrollIndicator={false}
        showsHorizontalScrollIndicator={false}
        automaticallyAdjustContentInsets={false}
        contentInsetAdjustmentBehavior="automatic"
        initialNumToRender={5}
        centerContent={true}
        contentContainerStyle={{ padding: 16, flexGrow: 1 }}
        renderItem={renderFeaturedprogram}
      />
    </Container>
  )

}

const styles = StyleSheet.create({
  safeAreaView: {
    flex: 1,
  },
  Main: {
    flex: 1,
  },
  programImageStyleBlock: {
    width: '90%',
    alignSelf: 'center'
  },
  programImageStyle: {
    width: '100%',
    height: 400,
  },
  loadingEmptyContainer: {
    width: '100%',
    height: '100%',
    backgroundColor: '#ffffff',
    alignItems: 'center',
    justifyContent: 'center',
  },
  loaderView: {
    width: 64,
    height: 64,
  },
  upnextText: {
    fontSize: 28,
    fontWeight: 'bold',
    marginVertical: 12,
    paddingHorizontal: 16,
  },
  programtitle: {
    width: "100%",
    fontSize: 18,
    fontWeight: 'bold',
    marginBottom: 3,
  },
  programtime: {
    width: "100%",
    fontSize: 16,
    color: "#636363",
    fontWeight: 'normal',
  },
  progWrapper: {
    width: "100%",
    flexDirection: 'row',
  },
  bars: {
    width: 16,
    height: 16,
    alignSelf: 'center',
    marginRight: 16,
  }
});

export default ProgramsScreen;
```

I really appreciate

**Reply by grahammendick** [reply]
You're not rendering your search results. You have to render your search results inside the `ScrollView` child of the `SearchBar`. This `ScrollView` must also have `contentInsetAdjustmentBehavior` set to 'automatic'.
```jsx
<SearchBar>
  <ScrollView contentInsetAdjustmentBehavior="automatic">
    <View>
      {productSearchData.map(() => )}
    </View>
  </ScrollView>
</SearchBar>
```

**Reply by grahammendick** [reply]
Did you try this ☝️ ?

**Reply by salisuabubakar** [reply]
@grahammendick  Yes i tried and it worked. Sorry for the late response. I had some issues. I would like to thank you once again for this wonderful library. Am in the process of submitting my first app with The Navigation Router. I can confidently say there is no difference between this and native app. I wish others will see this library and migrate because it's 100% native. I had a friend who doubted it but when i sent him the android apk file, he tested and told me he is migrating quickly. Thanks soo much for your time and please continue the good work.

**Reply by grahammendick** [reply]
Thank you so much. Telling a friend is the best donation you could ever give me. I've run into brick walls whenever I've tried to get the word out. I've given up trying.

You know, there isn't a single `View` component in the whole of the Navigation router because every component renders to the native primitive. The Navigation router tries to provide the exact same user experience you get if you were building an app without React Native. 

Anyway I'm really glad you're unblocked and the best of luck with your first app.

**Reply by salisuabubakar** [reply]
Hi @grahammendick ,
I could not get android to display the searchBar. Infact when i enable toolbar in the SearchBar android automatically crash. Kindly help me. Thanks

**Reply by grahammendick** [reply]
You have to set [the navigationImage prop on the NavigationBar](https://github.com/grahammendick/navigation/blob/8506784238ccaa0f6bb38a16af2018686ebc603b/NavigationReactNative/sample/zoom/Grid.js#L51). The Material 3 SearchBar expects one

**Reply by grahammendick** [reply]
@salisuabubakar You get it working?

**Reply by salisuabubakar** [reply]
Yes sorry for the late response. It’s working perfectly. Thanks

**Reply by grahammendick** [reply]
@salisuabubakar you don't have to set the `navigationImage` prop anymore. The [SearchBar uses the default one provided by Android instead](https://github.com/grahammendick/navigation/pull/710)
